### ISSUE TITLE
Speeding up checkout/build/operations (for maintenance releases)

### ISSUE THREAD

**Issue by jonathanolson** [phetsims/chipper#1112]
It's taking an impressively long time to do checkouts/builds (partly due to NPM). With the current process, if there is an emergency maintenance patch that needs to go out (where everything is broken), we would be mostly waiting on builds and checkouts to get the patch out, possibly in excess of 24-48 hours.

It seems best to minimize this, so that we could push emergency fixes out faster.

**Comment by jonathanolson** [comment]
> New idea: Set up a "maintenance optimized" checkout for each sim.

That sounds potentially quite useful! Would we expect devs doing a maintenance release to have that structure? I think this is definitely worthy of investigation.

**Comment by zepumph** [comment]
Would https://docs.github.com/en/actions/advanced-guides/caching-dependencies-to-speed-up-workflows be helpful?

**Comment by samreid** [comment]
We discussed this problem at today's dev meeting.

@jonathanolson: The proposal above would avoid the thrashing.  We would need up-front time to set up the checkouts with a script, and it would take more disk space.  We can look at the npm alternatives.  The network step and local is a burden.  Puppeteer requires reinstalling chrome which takes a long time especially when thrashing.
@zepumph: Would npm-cache help?
@jonathanolson: We have looked into it, but that gets bogged down and gives the 10-15 minutes problem until we clear the cache.
@jonathanolson I'd like to look at yarn.  Would the non-thrashing approach work well for puppeteer?
@zepumph: What if we get everything to the same level of puppeteer so it doesn't thrash?
@jonathanolson: Is that API stable?
@zepumph: It does seem so.  We haven't had struggles with it.
@samreid: Generally, we could try to get all node_modules on the same version?
@jonathanolson Maintenance releasing every sim would be less drastic than the non-thrashing approach.

@jonathanolson Who should be on the subteam, and how should we prioritize?  I'll take the lead and others can join as needed. 
@zepumph I can help if we try to unify the puppeteers.  At least testing if that's feasible.
@jonathanolson But we would also add puppeteer to sims that don't need it, so it doesn't have to clear and reinstall.  Any objections if we add puppeteer to all sims, even sims that don't need it.
@zepumph no objection, but if yarn just solves it maybe that's better.
@samreid: The non-thrashing would need to be implemented on the build server. 

@jonathanolson: I'll check out the npm costs and see what might help.  I can investigate the non-thrashing and yarn.  Upgrading versions and matching them so many sims have the same node_modules.

**Comment by pixelzoom** [comment]
Why is this in the "done" column of the Developer Meeting project board? It's still open.

**Comment by samreid** [comment]
In the "Investigate Deno" issue https://github.com/phetsims/chipper/issues/1272, we saw that Deno automatically manages dependencies without needing `npm install` or `npm update`.  That issue also proposes that we would get rid of `grunt` and `grunt-cli` so we wouldn't need `npm install` or `node_modules` in each sim repo.  But that step would require maintenance releasing every sim with new build tools "chipper 3.0" to make it happen.

**Comment by samreid** [comment]
@jonathanolson and I discussed this today, and talked about:

* Deno maintenance releases may be a reasonable long term vision, but would be prohibitively expensive to migrate everything to Deno in the short term
* For the "one checkout for each release branch" strategy, that still sounds plausible but would require significant changes to the maintenance release process.  That should run on developer machines (not just on bayes or equivalent).  Our next step will be to estimate the size for this strategy and see if it is prohibitively expensive.  Each sim would just check out its dependencies.  `node_modules` for chipper, perennial, and perennial-alias would probably take the most space.

**Comment by jonathanolson** [comment]
I'll work on parallel checkouts for the maintenance release process, we'll wait on the build-server based on how that goes (as it seems higher consequence and also involves a lot of people and coordination), and we'll both look into speeding up the build in the future.

**Comment by samreid** [comment]
@jonathanolson just finished several maintenance releases. Did you use parallel checkouts or other strategies for improving that process?  Any recommendations for changes for how other team members would do maintenance releases?

@marlitas and I are triaging subgroup issues, and it seems this can be "done discussing", but please move it back if desired. It is also labeled as an epic, so will likely be reviewed at a future quarterly planning meeting (so it won't be forgotten).

**Comment by jonathanolson** [comment]
Parallel checkouts are implemented, but the process isn't fully documented or fleshed out. I think I could use it for more parts (especially queries which check which release branches need a patch). I'd like to investigate that more.

For the build-server, https://github.com/phetsims/chipper/issues/1367 is probably the highest priority. It's possible parallel checkouts could speed up builds, I'd need to discuss with @mattpen.

**Comment by mattpen** [comment]
@jonathanolson - what are the expected storage requirements for parallel checkouts?

**Comment by jonathanolson** [comment]
> what are the expected storage requirements for parallel checkouts?

With all current release branches, I've got 71 GB. That number would grow in the future.

**Comment by mattpen** [comment]
phet-server2 has 799GB available, with 446GB more that can be allocated, so it sounds like it would be worth a try!  Should we set up a time in the first week of january to pair on this? Or is there some documentation I can follow?

**Comment by jonathanolson** [comment]
I'll update ReleaseBranch code as noted above in the commit, so that the maintenance tools and build-server can use parallel checkouts. I'll note that updateCheckout() will be used externally, and should be stable! Then I'll hand it to @mattpen for the remaining work.

Additionally, we discussed parallelizing builds, which could speed things up even further.

**Comment by samreid** [comment]
I experimented with a script to check out a sim's release branch and its dependencies. I thought this could be used to streamline the manual maintenance release process by having a separate checkout where all the SHAs and branches are already set up, and confirmed that the remote branch commit is correct.

Also, I felt this could be useful from an open-source nature, making it easy for someone to check out just one sim, and its dependencies, in a release branch.  

So I sketched out this grunt task:

<details>

```diff

Subject: [PATCH] Add documentation, see https://github.com/phetsims/chipper/issues/1369
---
Index: js/grunt/Gruntfile.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/grunt/Gruntfile.js b/js/grunt/Gruntfile.js
--- a/js/grunt/Gruntfile.js	(revision 60d6c6ea150cde378e6937c3372c9dc8f0dea235)
+++ b/js/grunt/Gruntfile.js	(date 1676668082637)
@@ -18,6 +18,7 @@
 const assert = require( 'assert' );
 const _ = require( 'lodash' ); // eslint-disable-line require-statement-match
 require( './checkNodeVersion' );
+
 ///////////////////////////
 
 module.exports = function( grunt ) {
@@ -92,6 +93,52 @@
       await checkoutDependencies( repo, dependencies, includeNpmUpdate );
     } ) );
 
+  grunt.registerTask( 'clone-target',
+    'Clone repos for a simulation and all of its declared dependencies\n' +
+    '--repo : repository name where package.json should be read from\n' +
+    '--target : the branch to check out\n' +
+    '--skipNpmUpdate : If provided, will prevent the usual npm update',
+    wrapTask( async () => {
+      const repo = grunt.option( 'repo' );
+      const target = grunt.option( 'target' );
+
+      assert( repo, 'Requires specifying a repository with --repo={{REPOSITORY}}' );
+      assert( target, 'Requires specifying a branch/SHA with --target={{BRANCH}}' );
+
+      assertIsValidRepoName( repo );
+
+      const cloneRepo = require( '../common/cloneRepo' );
+      const checkoutTarget = require( '../common/checkoutTarget' );
+      const gitCheckout = require( '../common/gitCheckout' );
+      const getDependencies = require( '../common/getDependencies' );
+      const execute = require( '../common/execute' );
+
+      await cloneRepo( repo );
+
+      // check out to the specified branch before getting dependencies, in case the list of dependencies in master differs
+      await gitCheckout( repo, target );
+
+      const dependenciesJSON = await getDependencies( repo );
+      const dependencyRepos = Object.keys( dependenciesJSON ).filter( dependency => dependency !== repo && dependency !== 'comment' );
+      for ( let i = 0; i < dependencyRepos.length; i++ ) {
+        await cloneRepo( dependencyRepos[ i ] );
+      }
+      await checkoutTarget( repo, target, !grunt.option( 'skipNpmUpdate' ) );
+
+      // Switch to branches as appropriate in case you want to commit:
+      // git ls-remote --heads origin branch-name
+      const branchName = `${repo}-${target}`;
+      for ( let i = 0; i < dependencyRepos.length; i++ ) {
+        const dependencyRepo = dependencyRepos[ i ];
+        const result = await execute( 'git', [ 'ls-remote', '--heads', 'origin', branchName ], `../${dependencyRepo}` ).trim();
+        if ( result.length > 0 ) {
+          const remoteSHA = result.split( ' \t' )[ 0 ];
+          console.log( dependencyRepo + ': ' + remoteSHA );
+        }
+      }
+    } )
+  );
+
   grunt.registerTask( 'checkout-target',
     'Check out a specific branch/SHA for a simulation and all of its declared dependencies\n' +
     '--repo : repository name where package.json should be read from\n' +

```

</details>

It's not complete at the end part--we would want to detect if the remote branch exists, and is at the right SHA.  Then check it out, or create a new local branch.

Anyways, I think it would take more time, so I don't have time to work on it in this iteration, but just wanted to write it down as an idea maybe for the future.  Basically, the steps to set up a parallel checkout would be:

```
mkdir gravity-and-orbits-1.6
cd gravity-and-orbits-1.6
git clone https://github.com/phetsims/perennial.git
cd perennial
npm install
grunt clone-target --repo=gravity-and-orbits --target=1.6
```

**Comment by zepumph** [comment]
This is very exciting! I'm going to note this for the next time I want to test another version of a sim. I'll let you know when I next come to this.

**Comment by samreid** [comment]
Thanks! I wanted to add a caveat that the first part of the patch above seems reasonable and seems to work, but I kind of ran out of steam/time at the end so the last part of the patch is very sketchy/untested.

**Comment by samreid** [comment]
Today @jessegreenberg and I discussed that this proposal would be a nice way to compare code changes between 2 releases (which we don't have any way of doing now).  This could help us be more confident in knowing how a maintenance RC differs from a prior publication or RC--and therefore better able to focus QA testing.

**Comment by mattpen** [comment]
Goals for "I" iteration: setup parallel checkouts for the build server on phet-server2.

**Comment by jonathanolson** [comment]
Looks like I had to make NPM be non-parallel. All tasks on my end handled with the above commit.

**Comment by mattpen** [comment]
@jonathanolson - It looks like build-server requests include the dependencies.json in the body.  This seems unneccessary, as the build-server can just fetch them from the release branch instead.  Are there some corner cases, like one-off builds, where the dependencies.json would be different from dependencies.json in the branch of the build request?

**Comment by mattpen** [comment]
> @jonathanolson - It looks like build-server requests include the dependencies.json in the body. This seems unneccessary, as the build-server can just fetch them from the release branch instead. Are there some corner cases, like one-off builds, where the dependencies.json would be different from dependencies.json in the branch of the build request?

It looks like this was in place because rosetta does not know about the branch name.  Currently all translations are failing due to lack of branch name.  @jonathanolson - can we find a time to collaborate on a solution for translations?  It looks like we'll need to put the branch name in the metadata or dependencies.json.

**Comment by zepumph** [comment]
Looks like an MR from JG resulted in an error for these release branches:


>After running through production deploys with the batch MR process, these new versions are not available.  Are these the wrong links or is there another issue?
https://phet-io.colorado.edu/sims/center-and-variability/1.0.11/center-and-variability_all_phet-io.html?phetioStandalone
https://phet-io.colorado.edu/sims/circuit-construction-kit-dc/1.3.5/circuit-construction-kit-dc_all_phet-io.html?phetioStandalone
https://phet-io.colorado.edu/sims/circuit-construction-kit-dc-virtual-lab/1.3.3/circuit-construction-kit-dc-virtual-lab_all_phet-io.html?phetioStandalone
https://phet-io.colorado.edu/sims/friction/1.6.5/friction_all_phet-io.html?phetioStandalone

And also ratio-and-proportion phet brand version 1.2.6

This is only critically broken for the phet-io links, as the RAP build just failed to build generally.

I reported:

Looks like this line is buggy:

https://github.com/phetsims/perennial/blob/a3257f46b7ad1e77742ac03742b7ae3658cf741a/js/build-server/taskWorker.js#L231-L230

Yielding a path like: ../release-branches/friction-1.6/friction/build/phet/phet-io
From the log:
>May 24 13:33:40 phet-server2.int.colorado.edu build-server[3953141]: debug: rsync: change_dir "/data/share/phet/build-server/perennial//../release-branches/friction-1.6/friction/build/phet/phet-io" failed: No such file or directory (2)
May 24 13:33:40 phet-server2.int.colorado.edu build-server[3953141]: debug: rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1189) [sender=3.1.3]

Bug introduced in https://github.com/phetsims/perennial/commit/a911b3f4a61e112837350aa5c23f3e45d57e6bde


Changed recently as part of https://github.com/phetsims/chipper/issues/1112


Talking with @mattpen and @chrisklus, they said they'll take it from there.

**Comment by chrisklus** [comment]
@mattpen and I fixed the buggy line identified above and tested by deploying CAV to phet-server-dev for phet AND phet-io brand. This worked as expected so we switched to production deploys.

All of the links from https://github.com/phetsims/chipper/issues/1112#issuecomment-1561919569 are fixed as well as RAP 1.2.6.

Leaving this issue open to continue work for a better long-term solution for translation builds, as described in https://github.com/phetsims/chipper/issues/1112#issuecomment-1561578444 (https://github.com/phetsims/perennial/commit/a3257f46b7ad1e77742ac03742b7ae3658cf741a was a temp solution).
